src/ParallelAlgorithms/LinearSolver/Multigrid/ElementActions.hpp

```diff
@@ -0,0 +1,415 @@
+ private:
+  using fields_tag = FieldsTag;
+  using residual_tag =
+      db::add_tag_prefix<LinearSolver::Tags::Residual, fields_tag>;
+
+ public:
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            size_t Dim, typename ActionList, typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&, Parallel::AlgorithmExecution,
+                    size_t>
+  apply(db::DataBox<DbTagsList>& box,
+        const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+        const Parallel::GlobalCache<Metavariables>& /*cache*/,
+        const ElementId<Dim>& /*element_id*/, const ActionList /*meta*/,
+        const ParallelComponent* const /*meta*/) noexcept {
+    const bool is_coarsest_grid =
+        not db::get<Tags::ParentId<Dim>>(box).has_value();
+
+    // Record pre-smoothing result fields and residual
+    if (db::get<Tags::OutputVolumeData<OptionsGroup>>(box)) {
+      db::mutate<Tags::VolumeDataForOutput<OptionsGroup, FieldsTag>>(
+          make_not_null(&box),
+          [](const auto volume_data, const auto& result_fields,
+             const auto& residuals) noexcept {
+            volume_data->assign_subset(
+                Variables<db::wrap_tags_in<Tags::PreSmoothingResult,
+                                           typename fields_tag::tags_list>>(
+                    result_fields));
+            volume_data->assign_subset(
+                Variables<db::wrap_tags_in<Tags::PreSmoothingResidual,
+                                           typename fields_tag::tags_list>>(
+                    residuals));
+          },
+          db::get<fields_tag>(box), db::get<residual_tag>(box));
+    }
+
+    // Skip the second smoothing step on the coarsest grid
+    const size_t first_action_after_post_smoothing_index = tmpl::index_of<
+        ActionList,
+        SendCorrectionToFinerGrid<FieldsTag, OptionsGroup, SourceTag>>::value;
+    const size_t this_action_index =
+        tmpl::index_of<ActionList, SkipPostsmoothingAtBottom>::value;
+    return {std::move(box), Parallel::AlgorithmExecution::Continue,
+            is_coarsest_grid ? first_action_after_post_smoothing_index
+                             : (this_action_index + 1)};
+  }
+};
+
+template <typename FieldsTag>
+struct CorrectionInboxTag
+    : public Parallel::InboxInserters::Value<CorrectionInboxTag<FieldsTag>> {
+  using temporal_id = size_t;
+  using type = std::map<temporal_id, typename FieldsTag::type>;
+};
+
+// The next two actions communicate and project the coarse-grid correction, i.e.
+// the solution of the post-smoother, to the finer grid. The post-smoother on
+// finer grids runs after receiving this coarse-grid correction. Since the
+// post-smoother is skipped on the coarsest level, it directly sends the
+// solution of the pre-smoother to the finer grid, thus kicking off the
+// "ascending" branch of the V-cycle.
+template <typename FieldsTag, typename OptionsGroup, typename SourceTag>
```

Comment:
The restriction actions are exposed publicly because an executable may need to restrict data from the finest grid down the multigrid hierarchy, e.g. variable background fields that get updated in every non-linear solver step. There's no need for executables to interpolate anything back up the multigrid hierarchy, so the prolongation operations are private to the multigrid algorithm and more tightly integrated into the algorithm.